### DISCUSSION TITLE
Request for feedback - Necklace

### DISCUSSION THREAD

**Discussion by johncclayton** [johncclayton/Necklace#3] [Ideas]
I began this with a post on my blob about the idea.

I'm looking for your feedback, is this a good idea?  Is it worth pursuing?  Would you use it?

For now I'm re-posting the content directly here, the original article: https://blog.effective-flow.ch/posts/necklace-appcast/

## Introduction

This is an idea - what I want to know is "would you use this?".  Keep reading to find out more - this could save thousands of developers thousands of hours of time. 
 
### Background 

I recently released an app on Windows - and I wanted my users to have a way to be notified that there was a new version of the app and make installation a breeze.  

Nothing new here - this has existed for years in the world of Mac apps, it's called [Sparkle](https://sparkle-project.org/), and on Windows I've used [NetSparkleUpdater](https://github.com/NetSparkleUpdater/NetSparkle).

### The Problem

The problem is not appcasts, they are an excellent way to encourage customers to update to the latest version - appcast libraries make this trivial to achieve _for the customer_.  

The problem is the __effort__ involved in getting it done.  Let me try to elaborate.

To make use of Sparkle - developers need to build the relevant client library into their code, and for each release they must:

  1. Have some storage space somewhere to publish assets to
  1. Publish (at least) both a binary asset and usually also a change log
  1. Re-generate the appcast.xml feed - this likely requires an appcast signing key
  1. Store the resulting binaries + change logs and appcast files on a web server / storage account. 
  1. Make the "latest" version available on their website 
  1. [Optional] List archived versions on their website

Depending on the library used; generation of the appcast might require the existing application assets to be available on disk.  This isn't always the case, but if it is then this can make the build pipelines not only less efficient - but you run the risk of turning your dev machine into a "build box", because you just happened to have all the binaries there.  

Then there is the process of setting up the app cast generation process itself - none of which is particularly difficult, there are lots of flavours I suppose they mostly go something like this: 

  1. Installation of the appcast generator on a machine (or in the build pipeline)
  1. Generation of signing keys - along with the responsibility to keep these safe
  1. Integration of the appcast generation process into your build process/pipeline
  1. Start thinking about having a beta channel, and not just stable... oh crap

### So what's my problem? 

I'm lazy - there is a lot I don't like about this process so far: 
  1. Most of this, if not all of it, is just technical infrastructure that developers need to repeat for each piece of software they ship.
  1. While the cusomters get "value" in the form of automated updates - its a complex set of tasks, there are lots of moving pieces to get it right: storage, tool installs, generation, channels and security. 
  1. It's boring, error prone work and could become a security hole - you are after all injecting code into your customers computer.  Should *anyone* be able to generate and ship an appcast?  Are those signing keys protected?
  1. Now you have to deal with hosting the appcast - and nothing I've read at the Sparkle site, nor on the NetSparkleUpdater site suggests this is "standard" in any way - great - [lets reinvent this again](https://en.wikipedia.org/wiki/Reinventing_the_wheel).

### The Vision / Proposal

So I'm asking myself - isn't this a ripe opportunity to make something better? 

What if it was possible to automatically create a production ready, secure appcast generation system in seconds?

What if this system made it childs play to expose different release channels, add releases to a channel, move a release to a different channel, and can automatically archive older versions of your published software?

If that system also provided a way to query for information via REST or GraphQL, then it would also be relatively simple to get a product web page populated with the latest build information - even with static site generators.

## Introducing: Necklace

Necklace is an open source project that provides developers with a ready to run appcast generation system.  

Necklace stores appcast meta data, handles separate channels for you, takes care of the generation as well as secure storage of your signing keys - and generates Sparkle/NetSparkleUpdater compatible appcast.xml files on-the-fly. 

There is no need to install app cast generator tools, copy/move/sync binaries or change logs or publish the resulting app cast. 

### What's it made of? 

Docker.  I'd like to make an Azure Functions version too - zero infra maintenace is a good thing, and since it wouldn't be called that often it would cost just cents.

Pull the image, configure it - expose it on a public HTTPS URL and *you are ready to publish releases*.

### How do I use it? 

Assuming you have a website hosted on a linux machine somewhere - to get up and running with Necklace, you would: 

1. Install the Docker based version of Necklace, you will want to bind mount a single directory to your local host in order to back up its configuration and database.  This is also where the signing keys will be exposed/set.
2. Set up the private appcast signing key configuration - the keys are automatically generated if they don’t exist within the directory that has been mapped.  You can simply overwrite them to use your own keys. 
3. Add your ChangeLog rendering template (a future feature)
4. Add configuration for the base hostname of the appcast binaries / packages. 

Then, when you have a new build to release to your customers (lets say it consists of a .pkg file and an .md change log file); you would then:

1. Store the .pkg / .md file somewhere accessible to Necklace - this could be GitHub, Azure Storage, AWS S3 and so on - you are probably *already doing this*
2. Make a REST call into Necklace to add the app version
3. There is no 3, you are DONE.

The new app version is now available to all your customers (on whatever channel you placed it).  If you re-publish the same version number in a different channel, the appcast system automatically reflects this.  

### Where do I get it? 

It doesn't exist yet - its just an idea, what I want to know is: is it attractive, shall I build this thing? 

Let me know.  Comments section below.

### Appcast Generator
 - [Sparkle Hub Appcast](https://github.com/marketplace/sparklehub-appcast)

### Things That Sparkle

 - [Sparkle](https://sparkle-project.org/)
 - [NetSparkleUpdater](https://github.com/NetSparkleUpdater/NetSparkle)
 - [WinSparkle](https://winsparkle.org/)

**Comment by dbarU** [comment]
Very good idea in my opinion. Although I don't have much to add to the discussion at the moment, just wanted to show my support for the idea/project. Your description sounds exactly like what I will have to be writing for my company's software provisioning capabilities.
